File: packages/digital-ink-recognition/android/src/main/java/io/capawesome/capacitorjs/plugins/mlkit/digitalinkrecognition/DigitalInkRecognition.java
package io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.google.android.gms.tasks.Task;
import com.google.mlkit.common.model.DownloadConditions;
import com.google.mlkit.common.model.RemoteModelManager;
import com.google.mlkit.vision.digitalink.common.RecognitionResult;
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognitionModel;
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognitionModelIdentifier;
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognizer;
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognizerOptions;
import com.google.mlkit.vision.digitalink.recognition.RecognitionContext;
import com.google.mlkit.vision.digitalink.recognition.WritingArea;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.CustomExceptions;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.options.DeleteDownloadedModelOptions;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.options.DownloadModelOptions;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.options.RecognizeOptions;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.results.GetDownloadedModelsResult;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.classes.results.RecognizeResult;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.mlkit.digitalinkrecognition.interfaces.NonEmptyResultCallback;

public class DigitalInkRecognition {

    @NonNull
    private final RemoteModelManager modelManager;

    public DigitalInkRecognition() {
        this.modelManager = RemoteModelManager.getInstance();
    }

    public void deleteDownloadedModel(@NonNull DeleteDownloadedModelOptions options, @NonNull EmptyCallback callback) throws Exception {
        DigitalInkRecognitionModel model = createModel(options.getLanguageTag());
        modelManager.deleteDownloadedModel(model).addOnSuccessListener(result -> callback.success()).addOnFailureListener(callback::error);
    }

    public void downloadModel(@NonNull DownloadModelOptions options, @NonNull EmptyCallback callback) throws Exception {
        DigitalInkRecognitionModel model = createModel(options.getLanguageTag());
        DownloadConditions conditions = new DownloadConditions.Builder().build();
        modelManager.download(model, conditions).addOnSuccessListener(result -> callback.success()).addOnFailureListener(callback::error);
    }

    public void getDownloadedModels(@NonNull NonEmptyResultCallback<GetDownloadedModelsResult> callback) {
        modelManager
            .getDownloadedModels(DigitalInkRecognitionModel.class)
            .addOnSuccessListener(models -> callback.success(new GetDownloadedModelsResult(models)))
            .addOnFailureListener(callback::error);
    }

    public void recognize(@NonNull RecognizeOptions options, @NonNull NonEmptyResultCallback<RecognizeResult> callback) throws Exception {
        DigitalInkRecognitionModel model = createModel(options.getLanguageTag());
        modelManager
            .isModelDownloaded(model)
            .addOnSuccessListener(isModelDownloaded -> {
                if (!isModelDownloaded) {
                    callback.error(CustomExceptions.MODEL_NOT_DOWNLOADED);
                    return;
                }
                DigitalInkRecognizerOptions recognizerOptions = DigitalInkRecognizerOptions.builder(model)
                    .setMaxResultCount(options.getMaxResultCount())
                    .build();
                DigitalInkRecognizer recognizer = com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognition.getClient(
                    recognizerOptions
                );
                RecognitionContext recognitionContext = createRecognitionContext(options.getPreContext(), options.getWritingArea());
                Task<RecognitionResult> task = recognitionContext == null
                    ? recognizer.recognize(options.getInk())
                    : recognizer.recognize(options.getInk(), recognitionContext);
                task
                    .addOnSuccessListener(recognitionResult -> {
                        recognizer.close();
                        callback.success(new RecognizeResult(recognitionResult));
                    })
                    .addOnFailureListener(exception -> {
                        recognizer.close();
                        callback.error(exception);
                    });
            })
            .addOnFailureListener(callback::error);
    }

    @NonNull
    private DigitalInkRecognitionModel createModel(@NonNull String languageTag) throws Exception {
        DigitalInkRecognitionModelIdentifier modelIdentifier;
        try {
            modelIdentifier = DigitalInkRecognitionModelIdentifier.fromLanguageTag(languageTag);
        } catch (Exception exception) {
            throw CustomExceptions.UNSUPPORTED_LANGUAGE_TAG;
        }
        if (modelIdentifier == null) {
            throw CustomExceptions.UNSUPPORTED_LANGUAGE_TAG;
        }
        return DigitalInkRecognitionModel.builder(modelIdentifier).build();
    }

    @Nullable
    private RecognitionContext createRecognitionContext(@Nullable String preContext, @Nullable WritingArea writingArea) {
        if (preContext == null && writingArea == null) {
            return null;
        }
        RecognitionContext.Builder builder = RecognitionContext.builder().setPreContext(preContext == null ? "" : preContext);
        if (writingArea != null) {
            builder.setWritingArea(writingArea);
        }
        return builder.build();
    }
}

File: packages/digital-ink-recognition/ios/Plugin/DigitalInkRecognition.swift
import Foundation
import MLKitDigitalInkRecognition

@objc public class DigitalInkRecognition: NSObject {
    private let plugin: DigitalInkRecognitionPlugin

    private var downloadQueue: [String: ((Error?) -> Void)] = [:]

    init(plugin: DigitalInkRecognitionPlugin) {
        self.plugin = plugin
    }

    @objc public func deleteDownloadedModel(_ options: DeleteDownloadedModelOptions, completion: @escaping (_ error: Error?) -> Void) throws {
        let model = try createModel(languageTag: options.languageTag)
        ModelManager.modelManager().deleteDownloadedModel(model) { error in
            completion(error)
        }
    }

    @objc public func downloadModel(_ options: DownloadModelOptions, completion: @escaping (_ error: Error?) -> Void) throws {
        let model = try createModel(languageTag: options.languageTag)
        if ModelManager.modelManager().isModelDownloaded(model) {
            completion(nil)
            return
        }
        let conditions = ModelDownloadConditions(
            allowsCellularAccess: true,
            allowsBackgroundDownloading: true
        )
        downloadQueue[model.modelIdentifier.languageTag] = completion
        ModelManager.modelManager().download(model, conditions: conditions)
    }

    @objc public func getDownloadedModels(completion: @escaping (_ result: GetDownloadedModelsResult?, _ error: Error?) -> Void) {
        DispatchQueue.global().async {
            var models = Set<DigitalInkRecognitionModel>()
            for modelIdentifier in DigitalInkRecognitionModelIdentifier.allModelIdentifiers() {
                let model = DigitalInkRecognitionModel(modelIdentifier: modelIdentifier)
                if ModelManager.modelManager().isModelDownloaded(model) {
                    models.insert(model)
                }
            }
            completion(GetDownloadedModelsResult(models: models), nil)
        }
    }

    public func handleModelDownloadDidFail(model: DigitalInkRecognitionModel, error: Error?) {
        let languageTag = model.modelIdentifier.languageTag
        if let completion = downloadQueue[languageTag] {
            downloadQueue[languageTag] = nil
            completion(error ?? CustomError.downloadModelFailed)
        }
    }

    public func handleModelDownloadDidSucceed(model: DigitalInkRecognitionModel) {
        let languageTag = model.modelIdentifier.languageTag
        if let completion = downloadQueue[languageTag] {
            downloadQueue[languageTag] = nil
            completion(nil)
        }
    }

    @objc public func recognize(_ options: RecognizeOptions, completion: @escaping (_ result: RecognizeResult?, _ error: Error?) -> Void) throws {
        let model = try createModel(languageTag: options.languageTag)
        if !ModelManager.modelManager().isModelDownloaded(model) {
            completion(nil, CustomError.modelNotDownloaded)
            return
        }
        let recognizerOptions = DigitalInkRecognizerOptions(model: model)
        recognizerOptions.maxResultCount = Int32(options.maxResultCount)
        let recognizer = DigitalInkRecognizer.digitalInkRecognizer(options: recognizerOptions)
        let handleRecognitionResult: (DigitalInkRecognitionResult?, Error?) -> Void = { recognitionResult, error in
            if let error = error {
                completion(nil, error)
                return
            }
            completion(RecognizeResult(candidates: recognitionResult?.candidates ?? []), nil)
        }
        if options.preContext == nil && options.writingArea == nil {
            recognizer.recognize(ink: options.ink, completion: handleRecognitionResult)
        } else {
            let context = DigitalInkRecognitionContext(preContext: options.preContext, writingArea: options.writingArea)
            recognizer.recognize(ink: options.ink, context: context, completion: handleRecognitionResult)
        }
    }

    private func createModel(languageTag: String) throws -> DigitalInkRecognitionModel {
        guard let modelIdentifier = try? DigitalInkRecognitionModelIdentifier.from(languageTag: languageTag) else {
            throw CustomError.unsupportedLanguageTag
        }
        return DigitalInkRecognitionModel(modelIdentifier: modelIdentifier)
    }
}
